src/liballoc/sync.rs

```diff
@@ -328,6 +328,40 @@ impl<T> Arc<T> {
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value.
+    #[inline]
+    #[unstable(feature = "arc_new_with", issue = "none")]
+    pub fn new_with(data_fn: impl FnOnce(&Weak<T>) -> T) -> Arc<T> {
```

Comment:
Since `Weak` is ptr-sized, isn't it odd to pass it by-reference here?

Reply:
This is what I previously did, but it requires I keep a second `Weak` around. If I pass ownership of the only `Weak` then `data_fn` could drop it, freeing the memory I'm about to write to. I've added this explanation in comments.